### PR TITLE
[DO NOT SQUASH] Add a rock-check-residency pass to check occupancy/residency of a given kernel

### PR DESCRIPTION
**Pull request by sjw36**
This pass check various metrics collected during rock lowering and AMDGPU backend captured in the ISA metadata.

Computes occupancy based on:

- Shared Buffer total size (rock lowering)
- VGPR allocation (amdgpu backend)
- SGPR allocation (amdgpu backend)

From these kernel requirements the rock-check-residency pass will compute blocks_per_cu occupancy within the arch target.

Implements this feature: https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/1248

### CODE REVIEW

**Comment by sjw36** [comment]
FYI @krzysz00 this may conflict with your upstream merge. If it is too much trouble it can wait until after the release.

**Comment by sjw36** [comment]
@krzysz00 @manupak @giuseros @ravil-mobile this is ready now. Please review.

File: mlir/lib/Dialect/Rock/Transforms/CheckResidency.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    if (!metaData)
+      return removeBinary();
+
+    GET_INT_VALUE2(nextVGPR, metaData, "amdhsa_next_free_vgpr");
+    GET_INT_VALUE2(nextSGPR, metaData, "amdhsa_next_free_sgpr");
+
+    LLVM_DEBUG(llvm::dbgs() << "  VGPR Size:          " << nextVGPR << "\n"
+                            << "  SGPR Size:          " << nextSGPR << "\n");
+
+    auto archAttr = mod->getParentOfType<ModuleOp>()->getAttrOfType<StringAttr>(
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/CheckResidency.cpp:86]: I know upstream they've added target specification attributes. Not directly relevant here but still

**Review by krzysz00 — APPROVED** [review]
Seems fine enough to me

**Pull request merged**